src/ast/nodes/ImportExpression.ts

```diff
@@ -45,6 +55,100 @@ export default class ImportExpression extends NodeBase {
 		this.source.bind();
 	}
 
+	/**
+	 * Get imported variables for deterministic usage, valid cases are:
+	 *
+	 * - `const { foo } = await import('bar')`.
+	 * - `(await import('bar')).foo`
+	 * - `import('bar').then(({ foo }) => {})`
+	 *
+	 * Returns empty array if it's side-effect only import.
+	 * Returns undefined if it's not fully deterministic.
+	 */
+	getDeterministicImportedNames(): readonly string[] | undefined {
+		const parent1 = this.parent;
+
+		// Side-effect only: import('bar')
+		if (parent1 instanceof ExpressionStatement) {
+			return EMPTY_ARRAY;
+		}
+
+		if (parent1 instanceof AwaitExpression) {
+			const parent2 = parent1.parent;
+
+			// Side-effect only: await import('bar')
+			if (parent2 instanceof ExpressionStatement) {
+				return EMPTY_ARRAY;
+			}
+
+			// Case 1: const { foo } = await import('bar')
+			if (parent2 instanceof VariableDeclarator) {
+				const declaration = parent2.id;
+				return declaration instanceof ObjectPattern
+					? getDeterministicObjectDestructure(declaration)
+					: undefined;
+			}
+
+			// Case 2: (await import('bar')).foo
+			if (parent2 instanceof MemberExpression) {
+				const id = parent2.property;
+				if (!parent2.computed && id instanceof Identifier) {
```

Comment:
It appears this `computed` case is the only missing test case, everything else is covered now. Good work!

Probably something like `await import('bar')[foo]`...

Reply:
We have the case already and the computed has already been covered - the missing coverage is the `id instanceof Identifier` - as when it's not computed, it seems it will only be able to be Identifier, we don't have such `(await import('bar')).0`. I think having this check would be good to be defensive but I guess we could also do a force type casting.